From bd973161e757038117c435ccf562ee0c5b8d7c0c Mon Sep 17 00:00:00 2001
From: Hongbo Yao <yaohongbo@huawei.com>
Date: Tue, 15 Nov 2022 23:03:59 +0800
Subject: [PATCH] x86/ioapic: add an interface for driver to sync hardware data
 into memory
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: bd973161e757038117c435ccf562ee0c5b8d7c0c
Modified-by-SEL: Yes, modified per the change in 341b4a72


hulk inclusion
category: bugfix
bugzilla: 20890, https://gitee.com/openeuler/kernel/issues/I5ZV1W

--------------------------------

The default interrupt trigger mode of Pci driver is Level, however,
some pci drivers want to change irq trigger mode to edge by writing
hardware register directly.
In current kernel, this kind of driver needs to sync entry data which
has already been reservd in memory.

This patch provides an interface for the driver to sync entry data in
memory.

Signed-off-by: Hongbo Yao <yaohongbo@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: zhangyi (F) <yi.zhang@huawei.com>
Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Zhang Jianhua <chris.zjh@huawei.com>
Reviewed-by: Liao Chang <liaochang1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/x86/include/asm/io_apic.h |  1 +
 arch/x86/kernel/apic/io_apic.c | 25 +++++++++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/arch/x86/include/asm/io_apic.h b/arch/x86/include/asm/io_apic.h
index a1a26f6d3aa4..294b99f1a84a 100644
--- a/arch/x86/include/asm/io_apic.h
+++ b/arch/x86/include/asm/io_apic.h
@@ -195,6 +195,7 @@ extern void clear_IO_APIC(void);
 extern void restore_boot_irq_mode(void);
 extern int IO_APIC_get_PCI_irq_vector(int bus, int devfn, int pin);
 extern void print_IO_APICs(void);
+extern void ioapic_sync_hardware_data(int irq);
 #else  /* !CONFIG_X86_IO_APIC */
 
 #define IO_APIC_IRQ(x)		0
diff --git a/arch/x86/kernel/apic/io_apic.c b/arch/x86/kernel/apic/io_apic.c
index 25b1d5c6af96..34bc1e3325a8 100644
--- a/arch/x86/kernel/apic/io_apic.c
+++ b/arch/x86/kernel/apic/io_apic.c
@@ -1201,6 +1201,31 @@ int IO_APIC_get_PCI_irq_vector(int bus, int slot, int pin)
 }
 EXPORT_SYMBOL(IO_APIC_get_PCI_irq_vector);
 
+/*
+ * The function is only provided for Drivers to keep the data value in
+ * memory consistent with the value of the register, and these drivers
+ * must use it carefully.
+ *
+ * The correct step should be:
+ * Change register ---> ioapic_sync_hardware_data ---> request_irq
+ */
+void ioapic_sync_hardware_data(int irq)
+{
+	struct IO_APIC_route_entry entry;
+	struct mp_chip_data *data;
+	struct irq_pin_list *pin_list;
+
+	data = irq_get_chip_data(irq);
+	for_each_irq_pin(pin_list, data->irq_2_pin) {
+		entry = ioapic_read_entry(pin_list->apic, pin_list->pin);
+		data->entry.is_level = entry.is_level;
+		data->entry.active_low= entry.active_low;
+		pr_debug("irq[%d] update trigger[%d] polarity[%d]\n",
+			irq, entry.is_level, entry.active_low);
+	}
+}
+EXPORT_SYMBOL(ioapic_sync_hardware_data);
+
 static struct irq_chip ioapic_chip, ioapic_ir_chip;
 
 static void __init setup_IO_APIC_irqs(void)
-- 
2.33.0

